tuple/postload.lua

```diff
@@ -4,7 +4,7 @@
 -- entry is removed and require is called once again).
 
 local ffi = require('ffi')
-local tuple_keydef = require('tuple.keydef')
+local tuple_keydef = ...
```

Comment:
I don't like this change. In lua modules it's common to expect that `(...) == module_name`, not the module itself.

See:

```console

tarantool> package.preload['foo'] = function(...) print(...) end
---
...

tarantool> require('foo')
foo

```

Would you like to use `local tuple_keydef = require(...)`?

Reply:
Or `package.loaded[...]`?

Reply:
`require('tuple.keydef')` does not work here (as well as `package.loaded['tuple.keydef']`), because `package.loaded['tuple.keydef']` is `nil` at the moment: `luaopen_tuple_keydef()` should return the table, then the interpreter will set it to `package.loaded`.

In theory we can set `package.loaded` manually, but we'll need to catch an error from postload.lua and set it back to `nil` in the case. I would go in another direction.

I can keep the name as the first argument and pass the module table as the second one.

Reply:
The change will look so:

```diff
diff --git a/tuple/keydef.c b/tuple/keydef.c
index 4e33de9..50c16dc 100644
--- a/tuple/keydef.c
+++ b/tuple/keydef.c
@@ -72,18 +72,24 @@ execute_postload_lua(struct lua_State *L)
 {
        int top = lua_gettop(L);
 
+       const char *modname = "tuple.keydef.postload";
        const char *modsrc = tuple_keydef_postload_lua;
        const char *modfile = "@tuple.keydef/postload.lua";
 
        if (luaL_loadbuffer(L, modsrc, strlen(modsrc), modfile) != 0)
                luaL_error(L, "Unable to load %s", modfile);
        /*
-        * Copy the module table to top of the Lua stack: it will
-        * be passed to the script and will be accessible as
-        * `...`.
+        * It is usual for modules to receive a module name as the
+        * first argument. Don't break this assumption.
         */
-       lua_pushvalue(L, -2);
-       lua_call(L, 1, 1);
+       lua_pushstring(L, modname);
+       /*
+        * Copy the module table to pass it to the script as the second
+        * argument. It will be accessible from the script as
+        * `select(2, ...)`.
+        */
+       lua_pushvalue(L, top);
+       lua_call(L, 2, 1);
 
        /* Ignore Lua return value. */
        lua_settop(L, top);
diff --git a/tuple/postload.lua b/tuple/postload.lua
index f262aa7..26a4a45 100644
--- a/tuple/postload.lua
+++ b/tuple/postload.lua
@@ -4,7 +4,7 @@
 -- entry is removed and require is called once again).
 
 local ffi = require('ffi')
-local tuple_keydef = ...
+local tuple_keydef = select(2, ...)
 local tuple_keydef_t = ffi.typeof('struct tuple_keydef')
 
 local methods = {
```

Does it look okay for you?

Reply:
BTW, I found that a qualified name is passed in the usual case. So, it should be 'tuple.keydef.postload', not just 'postload'.

Reply:
We discussed it with @rosik. Since this file is bundled into the .so / .dylib, Yaroslav is okay to don't simulate the usual module convention. He asked to highlight this in a comment at top of the postload.lua file.

Reply:
The postload.lua file header now looks so:

```lua
-- This script is NOT a regular module file.
--
-- It is bundled into the .so / .dylib file.
--
-- It is executed only on the first load of the module.
--
-- It is NOT executed after hot reload (when the package.loaded
-- entry is removed and require is called once again).
--
-- The tuple.keydef module table is accessible as `...`.
```

Reply:
Approve.